Add Apparmor profile

Talkless opened the pull request:
Introduce AppArmor profile increaced security in Linux, reducing simplest attack vectors.

AppArmor profiles where developed on various distros - Debian Strech and Debian Buster (Sid) with KDE/Gnome, Ubuntu 18.04, openSUSE Tumbleweed, and maybe more configurations I do not recall.

Closes #5484

- [ x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5577)
<!-- Reviewable:end -->

Talkless committed:
feat(apparmor): Add AppArmor profile

Introduce AppArmor profile, designed to work with AppArmor version
2.13.2 (Debian Buster).

Talkless committed:
feat(apparmor): Add AppArmor v2.12.1 profile

Copy 2.13.2 profile into 2.12.1 place to be the starting point for
modifying (backporting) AppArmor profile for version 2.12.1 (on Ubuntu
18.04, Debian Stretch, etc).

At this point 2.12.1 profile does not work, as AppArmor v2.12.1 does not
have needed abstractions and policy language features (such as `include
if exists`). Followup commits will fix these issues.

Talkless committed:
fix(apparmor): Fix AppArmor profile for version 2.12.1

* Remove `include if exists` usage.
* Remove @{uid} usage.
* Backport missing AppArmor abstractions as inline rules.

Talkless committed:
fix(apparmor): Fix loading libraries from custom install prefix

If qtox is installed in /usr/local prefix (for example), launching qTox
fails because loading libraries from @{qtox_prefix} directory was not
allowed.

Add rule to allow loading libraries from @{qtox_prefix}/lib directory.

Talkless committed:
fix(apparmor): Backport qt5 abstraction for v2.12.1 profile

AppArmor 2.12.1 does not have qt5 abstraction, with Qt5-related rules.

Backport qt5 abstraction from AppArmor upstream as inline rules.

Talkless committed:
fix(apparmor): Add ibus abstraction

IBus-related rules are needed detected on Gnome-based desktop (Ubuntu
18.40):
```
type=AVC msg=audit(1548508639.169:546): apparmor="DENIED"
operation="open" profile="qtox"
name="/home/vincas/.config/ibus/bus/c3d8689228fc49d8867d4e63e4408e23-unix-0"
pid=7653 comm="qtox" requested_mask="r" denied_mask="r" fsuid=1000
ouid=1000
```

Include ibus abstraction to fix IBus functionality.

Talkless committed:
fix(apparmor): Fix qTox cache access

AppAmor denies access to qTox cache directory:
```
type=AVC msg=audit(1548508759.153:640): apparmor="DENIED"
operation="mkdir" profile="qtox" name="/home/vincas/.cache/qTox/"
pid=7802 comm="qtox" requested_mask="c" denied_mask="c" fsuid=1000
ouid=1000
```

Add rule to allow access to qTox cache directory.

Talkless committed:
fix(apparmor): Fix accessibility DBus access

AppArmor denies access to a11y:
```
Jan 26 15:23:31 vincas-ubuntu1804 dbus-daemon: apparmor="DENIED"
operation="dbus_method_call"  bus="accessibility"
path="/org/freedesktop/DBus" interface="org.freedesktop.DBus"
member="Hello" mask="send" name="org.freedesktop.DBus" pid=8011
label="qtox" peer_label="unconfined"

Jan 26 15:23:31 vincas-ubuntu1804 dbus-daemon[1474]: apparmor="DENIED"
operation="dbus_method_call"  bus="session" path="/org/a11y/bus"
interface="org.freedesktop.DBus.Properties" member="Get" mask="send"
name="org.a11y.Bus" pid=8011 label="qtox" peer_pid=1620
peer_label="unconfined"
```

Include dbus-accessibility abstraction and one addition dbus rule to fix
denies.

Talkless committed:
fix(apparmor): Fix hunspell access

AppArmor denies access to hunspell files:
```
type=AVC msg=audit(1548511779.241:1773): apparmor="DENIED"
operation="open" profile="qtox" name="/usr/share/hunspell/lt_LT.aff"
pid=9833 comm="qtox" requested_mask="r" denied_mask="r" fsuid=1000
ouid=0

type=AVC msg=audit(1548511779.241:1774): apparmor="DENIED"
operation="open" profile="qtox" name="/usr/share/hunspell/lt_LT.dic"
pid=9833 comm="qtox" requested_mask="r" denied_mask="r" fsuid=1000
ouid=0
```

Add rule to allow reading hunspell dictionaries.

Talkless committed:
fix(apparmor): Fix dbus access

Add rules to allow DBus access (send & receive) to various DBus
interfaces. Detected on Ubuntu 18.04.

Talkless committed:
fix(apparmor): Fix access to openssl configuration

AppArmor denies access to openssl configuration files:
```
type=AVC msg=audit(1548516028.121:3031): apparmor="DENIED"
operation="open" profile="qtox" name="/etc/ssl/openssl.cnf" pid=12416
comm="qtox" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```

Fix deny by including openssl abstraction.

Talkless committed:
fix(apparmor): Fix .local/share/qTox/ access

Capturing desktop screenshot produces this DENIED messages:
```
type=AVC msg=audit(1548516170.837:3146): apparmor="DENIED"
operation="mkdir" profile="qtox" name="/home/vincas/.local/share/qTox/"
pid=12605 comm="qtox" requested_mask="c" denied_mask="c" fsuid=1000
ouid= 1000
```

Add rule to allow writing to .local/share/qTox/

Talkless committed:
fix(apparmor): Backport dri-enumerate abstraction

AppArmor 2.12 (Ubuntu 18.04) does not have dri-enumerate abstraction,
so policy compilation fails.

Backport dri-enumerate abstraction as inline rules.

Talkless committed:
fix(apparmor): Fix DBUS denies on Kubuntu 18.04

AppArmor denies access to systray on Kubuntu 18.04.

Add DBUS rules to make systray icon work.

Talkless committed:
fix(apparmor): Fix spam of DENIED messages on openSUSE

AppArmor produced spams lot's of log messages like these:
```
type=AVC msg=audit(1548784382.499:2192): apparmor="DENIED"
operation="file_mmap" profile="qtox" name="/tmp/#13317" pid=6389 comm="qtox"
requested_mask="m" denied_mask="m" fsuid=1000 ouid=1000
```

These appears to be libpcre2 mmaped shared memory, related to jitting.

Deny mmap()'ing files for execution from /tmp directory because currently there
is no way to allow shared memory access explicitly with AppArmor, so we choose
more secure way (while probably loosing regex performance).

Talkless committed:
fix(apparmor): backport kde abstraction

AppArmor upstream has new rules useful for running applications in KDE
desktop.

Backport rules from update kde abstraction to fix AppArmor denies.

Talkless committed:
fix(apparmor): fix file dialog on KDE desktop

Opening file dialog produces error:
```
type=AVC msg=audit(1549805942.022:1474): apparmor="DENIED"
operation="exec" profile="qtox"
name="/usr/lib/x86_64-linux-gnu/libexec/kf5/kioslave" pid=2784
comm="qtox" requested_mask="x" denied_mask="x" fsuid=1000 ouid=0
```

Add rule to allow launching kioslave helper.

Talkless committed:
fix(apparmor): fix file dialog denies

Add dbus and file rules to fix numerous denies when File Dialog is used
to select file for sending.

Talkless committed:
fix(apparmor): Fix font-related denies on openSUSE

Add file rules to fix numerous AppArmor denies related to fonts.

Talkless committed:
fix(apparmor): Fix openSUSE-related AppArmor denies

Add file rules for denies detected only in openSUSE desktop.

Talkless committed:
fix(apparmor): Backport fix from dri-common abstraction

AppArmor produces denies:
```
type=AVC msg=audit(1552817150.513:371): apparmor="DENIED"
operation="open" profile="qtox"
name="/usr/share/drirc.d/00-mesa-defaults.conf" pid=5895 comm="qtox"
requested_mas k="r" denied_mask="r" fsuid=1000 ouid=0
```

Fix is available in upstream.

Backport AppArmor commit 2d8d2f06d5697d9692330686bb5ddb0095621144 to fix
dri-related denies.

Talkless committed:
fix(apparmor): Fix screenshot capture under AppArmor

AppArmor denies creating screenshot file:
```
type=AVC msg=audit(1552817813.136:554): apparmor="DENIED"
operation="mkdir" profile="qtox"
name="/home/vincas/.local/share/Tox/qTox/images/" pid=16100 comm="qtox"
requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
```

Add file rule to allow creating desktop screenshot.

Talkless committed:
docs(apparmor): Add README about AppArmor

Document AppArmor profile usage, troubleshooting and other concerns.

Talkless committed:
feat(apparmor): Add AppArmor profile install scripts

Add scripts for installing AppArmor profiles.

Talkless committed:
docs(INSTALL.md): Add link to AppArmor readme.

Hint user about possibility to use AppArmor for security hardening in
ISNTALL.md.

Talkless committed:
docs(apparmor): Simplify install script usage

security/apparmor/install.sh scripts does not need to `cd` into it's
directory to work correctly.

Simplify example to invoke `install.sh` without `cd`.

Talkless committed:
docs(apparmor): Fix custom install prefix example

AppArmor profile attachement example is wrong because it does not use {
} "alternations" to make comma define alternative executable path.

Update example to make it actually work in custom install prefix case.

Talkless committed:
fix(apparmor): Fix typo in file path

File rule to allow loading the executable itself has typo - no
executable name itself is present. Profile still works OK but it might
fail on some older kernels.

Fix file rule by specifying full path to the executable.

sudden6 commented:
Do you have anything who can review the AppArmor part?

Talkless commented:
@sudden6 I'll try to find someone from AppArmor community.

setharnold commented:
This looks good; the profiles are pretty loose but should still provide some protection from a subverted process.

Thanks

Talkless commented:
@setharnold Loose you mean writing to (almost) whole home, or something else?

Talkless commented:
I will update 
```
  network tcp,
  network udp,
``` 
into more strict rules in coming days.

sudden6 reviewed (changes requested):
Looks good besides the tiny formatting issues.

Please also rebase onto latest master, else CI will cause problems.

> Reviewed 7 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5577#-:-LaHgepRBMdpJfBUPxs1:b-nhbih3)* status: 0 of 1 LGTMs obtained

___
*[security/apparmor/README.md, line 1 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHgB9qA4Qqv2VQYn0V:-LaHgB9qA4Qqv2VQYn0W:b1ou008) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L1)):*
> ```Markdown
> # Hardening qTox with AppArmor
> ```

Please reformat this file to 80 chars per line and move all links to the bottom like in README.md. (Links do not need to adhere to 80 chars)
___
*[security/apparmor/README.md, line 11 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHf3wb-GqDkwIVdIpl:-LaHf3wc-1SZWhdEU1S8:b-orjtqr) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L11)):*
> ```Markdown
> Consider using additional security measures like [Firejail](https://firejail.wordpress.com/) to improve security even more.
> 
> Please also note that not all distributions has full AppArmor feature set available. For example, Debian (at least up to Debian 10 (buster)) does not have network, DBus mediation available. Also, X Server, shared user configuration files (like `~/.config/QtProject.conf`, caches, etc), opening web links via unconfined browsers introduces additional attack vectors, too. So please be cautious even with number of security measures applied.
> ```

`all distributions have`
___
*[security/apparmor/README.md, line 54 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHfOoO5hRCTFipbEkc:-LaHfOoO5hRCTFipbEkd:bef9vxw) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L54)):*
> ```Markdown
> ```
> 
> If OK it's marked as `(enforce)`. `uncofined` means AppArmor profile is not attached to the process, no confinement is being applied.
> ```

typo `unconfined`


<!-- Sent from Reviewable.io -->

sudden6 commented:
@setharnold thank you for taking a look

Talkless committed:
fix(apparmor): Make network rules more strict

Explicitly define allowed network domain.

Talkless committed:
docs(apparmor): Refactor links

Move links to the bottom of the document to conform to the qTox
documentation guidelines.

Talkless committed:
docs(apparmor): Fix line lengths

Wrap text to 80 chars per line to conform to qTox documentation
guidelines.

Talkless committed:
docs(apparmor): Fix grammar

Fix documentation text errors detected in review.

Talkless reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5577)* status: 0 of 1 LGTMs obtained (waiting on @sudden6)

___
*[security/apparmor/README.md, line 1 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHgB9qA4Qqv2VQYn0V:-Laq2bK6Ds3U0CPTABlE:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L1)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Please reformat this file to 80 chars per line and move all links to the bottom like in README.md. (Links do not need to adhere to 80 chars)
</blockquote></details>

Done.
___
*[security/apparmor/README.md, line 11 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHf3wb-GqDkwIVdIpl:-Laq2d0Y968I_oncXefm:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L11)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

`all distributions have`
</blockquote></details>

Done.
___
*[security/apparmor/README.md, line 54 at r1](https://reviewable.io/reviews/qtox/qtox/5577#-LaHfOoO5hRCTFipbEkc:-Laq2e_GCnEPWE-S3l9T:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/87d734e9564f188019e8409127e05eb30455dd99/security/apparmor/README.md#L54)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

typo `unconfined`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Talkless reviewed (commented):
One commit `fix(apparmor): backport kde abstraction` stars with lovercase `b`, should I fix that @sudden6 ?

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5577#-:-Laq3Kr3E3gHcYr3QAm8:b-8z69c3)* status: 0 of 1 LGTMs obtained (waiting on @sudden6)



<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
> One commit fix(apparmor): backport kde abstraction stars with lovercase b, should I fix that @sudden6 ?
No, that's not a problem.

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5577#-:-LaqGD_k1yt6a1nSro8X:bfeeb7q)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->
